feature/amazon/src/main/java/com/revenuecat/purchases/amazon/AmazonBilling.kt

```diff
@@ -140,31 +187,187 @@ class AmazonBilling constructor(
+        }
+    }
+
+    private fun getMissingSkusForReceipts(
+        appUserID: String,
+        amazonUserID: String,
+        receipts: List<Receipt>,
+        onCompletion: (tokensToSkusMap: Map<String, String>, errors: Map<String, PurchasesError>) -> Unit
+    ) {
+        val currentlyCachedTokensToSkus: Map<String, String> = cache.getReceiptSkus()
+
+        val successMap: MutableMap<String, String> = currentlyCachedTokensToSkus.toMutableMap()
+        val errorMap: MutableMap<String, PurchasesError> = mutableMapOf()
+
+        val nonSubscriptionReceiptsToSku =
+            receipts.filterNot { it.productType == ProductType.SUBSCRIPTION }
+                .map { it.receiptId to it.sku }
+
+        successMap.putAll(nonSubscriptionReceiptsToSku)
+
+        val subscriptionReceiptsToFetchTermSku: List<Receipt> =
+            receipts
+                .filter { it.productType == ProductType.SUBSCRIPTION }
+                .filterNot { currentlyCachedTokensToSkus.containsKey(it.receiptId) }
+
+        if (subscriptionReceiptsToFetchTermSku.isEmpty()) {
+            onCompletion(successMap, errorMap)
+            return
+        }
+
+        var receiptsLeft = subscriptionReceiptsToFetchTermSku.count()
+        subscriptionReceiptsToFetchTermSku.forEach { receipt ->
+            amazonBackend.getAmazonReceiptData(
+                receipt.receiptId,
+                amazonUserID,
+                onSuccess = { response ->
+                    log(LogIntent.DEBUG, AmazonStrings.RECEIPT_DATA_RECEIVED.format(response.toString()))
+
+                    successMap[receipt.receiptId] = response["termSku"] as String
+
+                    receiptsLeft--
+                    if (receiptsLeft == 0) {
+                        cache.setReceiptSkus(successMap)
+                        onCompletion(successMap, errorMap)
+                    }
+                }
+            ) { error ->
+                log(LogIntent.AMAZON_ERROR, AmazonStrings.ERROR_FETCHING_RECEIPT_INFO.format(error))
+
+                errorMap[receipt.receiptId] = error
+
+                receiptsLeft--
+                if (receiptsLeft == 0) {
+                    onCompletion(successMap, errorMap)
+                }
+            }
+        }
+    }
+
     private fun handleReceipt(
         receipt: Receipt,
-        appUserID: String,
         userData: UserData,
         productDetails: ProductDetails,
         presentedOfferingIdentifier: String?
     ) {
+        if (receipt.productType != ProductType.SUBSCRIPTION) {
+            val amazonPurchaseWrapper = AmazonPurchaseWrapper(
+                sku = productDetails.sku,
+                containedReceipt = receipt,
+                presentedOfferingIdentifier = presentedOfferingIdentifier,
+                purchaseState = RevenueCatPurchaseState.PURCHASED,
+                storeUserID = userData.userId
+            )
+            purchasesUpdatedListener?.onPurchasesUpdated(listOf(amazonPurchaseWrapper))
+            return
```

Comment:
if the product is not a subscription, we don't need the termSku, we can just use the sku

Reply:
maybe add that as a comment?

Reply:
Added in 0734edd